File: frag-nav/src/main/java/com/ncapdevi/fragnav/tabhistory/CollectionFragNavTabHistoryController.java
package com.ncapdevi.fragnav.tabhistory;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.ncapdevi.fragnav.FragNavPopController;
import com.ncapdevi.fragnav.FragNavSwitchController;
import com.ncapdevi.fragnav.FragNavTransactionOptions;

import java.util.ArrayList;

abstract class CollectionFragNavTabHistoryController extends BaseFragNavTabHistoryController {
    private static final String EXTRA_STACK_HISTORY = "EXTRA_STACK_HISTORY";

    private FragNavSwitchController fragNavSwitchController;

    CollectionFragNavTabHistoryController(FragNavPopController fragNavPopController,
                                          FragNavSwitchController fragNavSwitchController) {
        super(fragNavPopController);
        this.fragNavSwitchController = fragNavSwitchController;
    }

    @Override
    public boolean popFragments(int popDepth,
                                FragNavTransactionOptions transactionOptions) throws UnsupportedOperationException {
        boolean changed = false;
        boolean switched;
        do {
            switched = false;
            int count = fragNavPopController.tryPopFragments(popDepth, transactionOptions);
            if (count > 0) {
                changed = true;
                switched = true;
                popDepth -= count;
            } else if (getCollectionSize() > 1) {
                fragNavSwitchController.switchTab(getAndRemoveIndex(), transactionOptions);
                popDepth--;
                changed = true;
                switched = true;
            }
        } while (popDepth > 0 && switched);
        return changed;
    }

    abstract int getCollectionSize();

    abstract int getAndRemoveIndex();

    @NonNull
    abstract ArrayList<Integer> getHistory();

    abstract void setHistory(@NonNull ArrayList<Integer> history);

    @Override
    public void restoreFromBundle(@Nullable Bundle savedInstanceState) {
        if (savedInstanceState == null) {
            return;
        }
        ArrayList<Integer> arrayList = savedInstanceState.getIntegerArrayList(EXTRA_STACK_HISTORY);
        if (arrayList != null) {
            setHistory(arrayList);
        }
    }

    @Override
    public void onSaveInstanceState(@NonNull Bundle outState) {
        ArrayList<Integer> history = getHistory();
        if (history.isEmpty()) {
            return;
        }
        outState.putIntegerArrayList(EXTRA_STACK_HISTORY, history);
    }
}

File: frag-nav/src/main/java/com/ncapdevi/fragnav/tabhistory/CollectionFragNavTabHistoryController.kt
package com.ncapdevi.fragnav.tabhistory

import android.os.Bundle
import com.ncapdevi.fragnav.FragNavPopController
import com.ncapdevi.fragnav.FragNavSwitchController
import com.ncapdevi.fragnav.FragNavTransactionOptions
import java.util.*

abstract class CollectionFragNavTabHistoryController(fragNavPopController: FragNavPopController,
                                                     private val fragNavSwitchController: FragNavSwitchController) : BaseFragNavTabHistoryController(fragNavPopController) {

    internal abstract val collectionSize: Int

    internal abstract val andRemoveIndex: Int

    internal abstract var history: ArrayList<Int>

    @Throws(UnsupportedOperationException::class)
    override fun popFragments(popDepth: Int,
                              transactionOptions: FragNavTransactionOptions?): Boolean {
        var localDepth = popDepth
        var changed = false
        var switched: Boolean
        do {
            switched = false
            val count = fragNavPopController.tryPopFragments(localDepth, transactionOptions)
            if (count > 0) {
                changed = true
                switched = true
                localDepth -= count
            } else if (collectionSize > 1) {
                fragNavSwitchController.switchTab(andRemoveIndex, transactionOptions)
                localDepth--
                changed = true
                switched = true
            }
        } while (localDepth > 0 && switched)
        return changed
    }

    override fun restoreFromBundle(savedInstanceState: Bundle?) {
        if (savedInstanceState == null) {
            return
        }
        val arrayList = savedInstanceState.getIntegerArrayList(EXTRA_STACK_HISTORY)
        if (arrayList != null) {
            history = arrayList
        }
    }

    override fun onSaveInstanceState(outState: Bundle) {
        val history = history
        if (history.isEmpty()) {
            return
        }
        outState.putIntegerArrayList(EXTRA_STACK_HISTORY, history)
    }

    companion object {
        private const val EXTRA_STACK_HISTORY = "EXTRA_STACK_HISTORY"
    }
}
